docs/static-module-expansion.md

```diff
@@ -0,0 +1,277 @@
+# Static Module Expansion
+
+Modules may be expanded using for_each and count. This poses a problem for the static evaluation step.
+
+For example:
+```hcl
+# main.tf
+module "mod" {
+        for_each = {"us" = "first", "eu" = "second"}
+        source = "./mod"
+        name = each.key
+        providers {
+          aws = provider.aws[each.value]
+        }
+}
+```
+
+Each "instance" of "mod" will need it's own provider configuration, which is currently only specified as part of "mod" and not any "mod instance". This problem becomes more complex and problematic when considering nested modules.
+
+To solve this, modules which have static for_each and count expressions must be expanded at the config layer. This must happen before the graph building, transformers, and walking.
```

Comment:
Maybe I do not understand the meaning behind this, but I don't think we can **always** expand the module at config layer. We still have to support cases where the `for_each` is only evaluated in plan walk

Reply:
That's sort of what I meant by "static for_each and count expressions", but could reword it as "for_each and count expressions that can be statically evaluated"